### PR TITLE
Add POW confinement validation

### PR DESCRIPTION
**Pull request by cvalarida**
## Description
Adds a validation to the POW confinement fields to ensure that each confinement period is within a single period of service.

## Testing done
Unit testing for the validation function, manual testing to ensure it looks like we expect it to.

## Screenshots
Before entering data into the last field
image

After entering data into the last field
image

## Acceptance criteria
- [x] The user can't continue past the page if a period of confinement isn't completely within a period of service

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by cvalarida** [comment]
@peggygannon What should the error message say?

@goldenmeanie The UX is a bit wonky, but I'm not sure how to fix it. I can't add a single validation message that spans both the `to` and `from` dates, so I have to add them to each individually. Should we do something else instead?

File: src/applications/disability-benefits/all-claims/utils.jsx [hunk 1]
```diff
@@ -537,3 +537,37 @@ export const getAttachmentsSchema = defaultAttachmentId => {
...
+ * @property {string} to - A date string YYYY-MM-DD
+ * @property {string} from - A date string YYYY-MM-DD
+ * ---
+ * @param {String|DateRange} inside - The date or date range to check
+ * @param {DateRange} outside - The range `inside` must fit in
+ * @param {String} inclusivity - See https://momentjs.com/docs/#/query/is-between/
+ *                               NOTE: This function defaults to inclusive dates which is different
+ *                               from moment's default
+ */
+export const isWithinRange = (inside, outside, inclusivity = '[]') => {
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/utils.jsx]: had a little bit of trouble parsing this based on the param names, but should've probably read the jsdoc 😛

File: src/applications/disability-benefits/all-claims/utils.jsx [hunk 2]
```diff
@@ -537,3 +537,37 @@ export const getAttachmentsSchema = defaultAttachmentId => {
...
 };
+
+const isDateRange = ({ from, to }) => !!(from && to);
+
+const parseDate = dateString => moment(dateString, 'YYYY-MM-DD');
+
+// NOTE: Could move this to outside all-claims
+/**
+ * Checks to see if the first parameter is inside the date range (second parameter).
+ * If the first parameter is a date range, it'll return true if both dates are inside the range.
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/utils.jsx]: this is nifty, I like it

File: src/applications/disability-benefits/all-claims/validations.js [hunk 3]
```diff
@@ -205,3 +206,19 @@ export const hasMonthYear = (err, fieldData) => {
     err.addError('Please provide both month and year');
   }
 };
+
+export const isWithinServicePeriod = (errors, fieldData, formData) => {
+  const inServicePeriod = _.get(
+    'serviceInformation.servicePeriods',
+    formData,
+    [],
+  ).some(pos => isWithinRange(fieldData, pos.dateRange));
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/validations.js]: 💯

File: src/applications/disability-benefits/all-claims/validations.js [hunk 4]
```diff
@@ -205,3 +206,19 @@ export const hasMonthYear = (err, fieldData) => {
...
+  if (!inServicePeriod) {
+    const dateIsComplete = dateString =>
+      dateString && !dateString.includes('X');
+    if (dateIsComplete(fieldData.from) && dateIsComplete(fieldData.to)) {
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/validations.js]: We don't allow partial entry for POW, right? So this basically waits until the user is finished inputting the dates before the sirens start up?

File: src/applications/disability-benefits/all-claims/validations.js [hunk 5]
```diff
@@ -205,3 +206,19 @@ export const hasMonthYear = (err, fieldData) => {
...
+      errors.from.addError('Must be within a single period of service.');
+      errors.to.addError('Must be within a single period of service.');
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/validations.js]: 🤔 yeah not ideal to have the error show in both spots as opposed to once at the top of the date range UI.
Wonder what we can do about this.

File: src/applications/disability-benefits/all-claims/tests/pages/prisonerOfWar.unit.spec.jsx [hunk 6]
```diff
@@ -16,13 +16,19 @@ describe('Prisoner of war info', () => {
     uiSchema,
   } = formConfig.chapters.disabilities.pages.prisonerOfWar;
 
+  const formData = {
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/tests/pages/prisonerOfWar.unit.spec.jsx]: I realize that you're testing the `isWithinRange` function separately, but feels as though we should also maybe add a couple of tests into the POW page to ensure that the validation is being applied to the UI correctly. I.e., we get error(s) on the page, and can't continue.

**Review by markgreenburg — APPROVED** [review]
LGTM! Couple of things:
- Not sure what to do about the double error message, but in lieu of a better solution I think this is the best we can do.
- Could probably use some extra unit tests for the POW page given the new validation

**Comment by peggygannon** [comment]
@cvalarida @markgreenburg @goldenmeanie 
Let me know if you think this would work instead of the double error messaging:  

Please tell us your dates of confinement. The dates need to be within a single service period. 
From (*Required)
{date}

To (*Required)
{date}

**Review by bshyong — APPROVED** [review]

File: src/applications/disability-benefits/all-claims/utils.jsx (hunk 1 above)
Comment (cvalarida) [reply src/applications/disability-benefits/all-claims/utils.jsx]: Yeahhhhhh I don't like the naming. Any suggestions?

File: src/applications/disability-benefits/all-claims/validations.js (hunk 4 above)
Comment (cvalarida) [reply src/applications/disability-benefits/all-claims/validations.js]: Right, yeah. I think they have to enter the whole date.

That's exactly it.

File: src/applications/disability-benefits/all-claims/validations.js (hunk 5 above)
Comment (cvalarida) [reply src/applications/disability-benefits/all-claims/validations.js]: Change the way validations work in USFS' `ObjectField`. :trollface:

**Comment by cvalarida** [comment]
@peggygannon I like telling the user up front. What should the validation message say if they aren't within a single service period, though?

**Comment by peggygannon** [comment]
Hi, @cvalarida here's my suggested  validation error: 

The dates you enter must be within a single service period.

**Comment by cvalarida** [comment]
Perfect. Thanks, @peggygannon! :raised_hands:

**Comment by cvalarida** [comment]
So this is what I've got now. What do you think, @goldenmeanie?
image

Alternatively, we could remove the second error message and just highlight it red:
image

**Comment by goldenmeanie** [comment]
Let's chat about this @cvalarida

**Comment by goldenmeanie** [comment]
This would be nice:
screen shot 2019-01-24 at 3 41 56 pm

This seems easier:
screen shot 2019-01-24 at 3 41 47 pm

Thanks for the chat @cvalarida!

**Pull request merged**